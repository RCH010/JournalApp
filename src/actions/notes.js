import Swal from 'sweetalert2';
import { db } from "../firebase/firebaseConfig";
import { fileUpload } from '../helpers/fileUpload';
import { loadNotes } from "../helpers/loadNotes";
import { types } from "../types/types";


// con lo de thunk, que es lo que ayuda con el middleware para las acciones asincronas
//  tiene además del dispatch, el getState, y se puede acceder a esos, y ya del
//  get state, podemos acceder a los valores del state, o con el dispatch para
//  ejecutar una acción y se modifique algo del state
export const startNewNote = () => {
    return async (dispatch, getState) => {
        const {uid} = getState().auth;
        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const docRef = await db.collection(`${uid}/journal/notes`).add(newNote);
        dispatch(activeNote(docRef.id, newNote));
        dispatch(addNewNote(docRef.id, newNote));
    }
}
// Action to set a note in the state
export const activeNote = (id, note) => (
    {
        type: types.notesActive,
        payload: {
            id,
            ...note
        }
    }
)

export const addNewNote = (id, note) => ({
    type: types.notesAddNew,
    payload: {
        id,
        ...note
    }
})

export const startLoadingNotes = (uid) => {
    return async (dispatch) => {
        const notes = await loadNotes(uid);
        dispatch(setNotes(notes));
    }
}
 
export const setNotes = (notes) => ({
    type: types.noteLoad,
    payload: notes
});

// action to update notes in firestore
export const startSaveNote = (note) => {
    return async (dispatch, getState) => {
        const {uid} = getState().auth;

        
        const noteToFirestore = {...note};
        delete noteToFirestore.id;

        if(!noteToFirestore.url) { delete noteToFirestore.url; }
        if(!noteToFirestore.body) { delete noteToFirestore.body; }

        await db.doc(`/${uid}/journal/notes/${note.id}`).update(noteToFirestore)
            .then(() => {
                dispatch(refreshNote(note.id, note));
                Swal.fire('Saved', note.title, 'success');
            })
            .catch(() => {
                Swal.fire('Something went wrong :(', note.title, 'error');
            })
    }
}

// action to update the note that's just been updated
//  so in the list of notes it is updated
export const refreshNote = (id, note) => ({
    type: types.notesUpdated,
    payload: {
        id, 
        note: {
            id,
            ...note
        }
    }
})

// action that calls ./helpers/fileUpload to upload the image and returns
//  image url, then we call action startSaveNote so the db is updated
export const startUploadingPicture = (file) => {
    return async (dispatch, getState) => {
        const {active: activeNote} = getState().notes;
        // UI feedback
        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait.',
            allowOutsideClick: false,
            showConfirmButton: false,
            willOpen: () => {
                Swal.showLoading();
            }
        }); 
        const fileUrl = await fileUpload(file);
        
        activeNote.url = fileUrl;
        dispatch(startSaveNote(activeNote));

        Swal.close();
    }
}

//
export const startDeleting = (noteId) => {
    return async (dispatch, getState) => {
        const {uid} = getState().auth;
        // deleted from firestore
        await db.doc(`${uid}/journal/notes/${noteId}`).delete();

        //now deleted from store
        dispatch(deleteNote(noteId));
    }
}

export const deleteNote = (noteId) => ({
    type: types.notesDelete,
    payload: noteId,
})

export const notesLogout = () => ({
    type: types.notesLogoutCleaning,
})
