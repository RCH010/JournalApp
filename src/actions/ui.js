import { types } from "../types/types"

export const setError = (errorMsg) => {
    return {
        type: types.uiSetError,
        payload: errorMsg
    };
};

export const removeError = () => {
    return {
        type: types.uiRemoveError
    };
};

export const startLoading = () => {
    return {
        type: types.uiStartLoading
    }
};

export const finishLoading = () => {
    return {
        type: types.uiFinishLoading
    };
}