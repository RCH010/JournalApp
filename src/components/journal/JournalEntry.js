import React from 'react'

export const JournalEntry = () => {
    return (
        <div className='journal__entry pointer'>
            <div 
                className='journal__entry-picture'
                style={{
                    backgroundSize: 'cover',
                    backgroundImage: 'url(https://www.vets4pets.com/siteassets/species/cat/kitten/tiny-kitten-in-sunlight.jpg?w=585&scale=down)'
                }}
            >
            </div>
            <div className='journal__entry-body'>
                <p className='journal__entry-title'>
                    Un nuevo d√≠a
                </p>
                <p className='journal__entry-content'>
                    Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
                </p>
            </div>
            <div className='journal__entry-date-box'>
                <span>Monday</span>
                <h4>28</h4>
            </div>
        </div>
    )
}