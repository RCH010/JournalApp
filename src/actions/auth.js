import { googleAuthProvider, firebase } from "../firebase/firebaseConfig";
import { types } from "../types/types"
import { finishLoading, startLoading } from "./ui";

//  CUANDO ES UNA TAREA ASINCRONA, SE NECESITA RETORNAR UN CALLBACK
export const startLoginWithEmailPassword = (email, password) => {
    return (dispatch) => {
        dispatch(startLoading());
        firebase.auth().signInWithEmailAndPassword(email,password)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName, user.email)
                );
            })
            .catch(e => {
                console.log(e);
            })
            .finally(() => {
                dispatch(finishLoading())
            })
    }
}

export const startRegisterWithEmailPasswordName = (email, password, name) => {
    return (dispatch) => {
        firebase.auth().createUserWithEmailAndPassword(email,password)
            .then( async({user}) => {
                await user.updateProfile({displayName: name});
                dispatch(
                    login(user.uid, user.displayName, user.email)
                )
            })
            .catch( e => {
                console.log(e);
            })
    }
}

export const startGoogleLogin = () => {
    return (dispatch) => {
        firebase.auth().signInWithPopup(googleAuthProvider)
            .then(({user}) => {
                dispatch(
                    login(user.uid, user.displayName, user.email)
                )
            });
    }
}

export const login = (uid, displayName, email) => ( {
    type: types.login,
    payload: {
        uid,
        displayName,
        email
    }
})