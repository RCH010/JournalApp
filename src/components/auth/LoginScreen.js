import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import validator from 'validator'
import { startGoogleLogin, startLoginWithEmailPassword } from '../../actions/auth'
import { removeError, setError } from '../../actions/ui'
import { useForm } from '../../hooks/useForm'

export const LoginScreen = () => {
    
    const dispatch = useDispatch()
    const {msgError, loading} = useSelector(state => state.ui);

    const [formValues, handleInputChange] = useForm({
        email: '',
        password: '',
    })

    const { email, password } = formValues;

    const handleLogin = (e) => {
        e.preventDefault();
        if(isFormValid()){
            dispatch(startLoginWithEmailPassword(email, password));
        }

    }

    const handleGoogleLogin = () => {
        dispatch(startGoogleLogin());
    }

    const isFormValid = () => {
        if(!validator.isEmail(email)) {
            dispatch(setError('Email is invalid.'));
            return false;
        }
        if(password.length <= 5) {
            dispatch(setError('Password should be at least 6 characters.'));
            return false;
        }
        dispatch(removeError());
        return true;
    }

    return (
        <div>
            <h3 className='auth__title'>Login</h3>

            <form onSubmit={handleLogin}>
                {
                    msgError && 
                    (<div className='auth__alert-error'>
                        {msgError}
                    </div>)
                }
                <input 
                    type='email'
                    placeholder='myjournal@gmail.com'
                    name='email'
                    className='auth__input'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />
                <input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    className='auth__input'
                    value={password}
                    onChange={handleInputChange}
                />

                <button
                    type='submit'
                    className='btn btn-primary btn-block'
                    disabled={loading}
                >
                    Login
                </button>
                

                <div className='auth__social-networks'>
                    <p className='auth__social-networks-title mt-1'>Login with social networks</p>
                
                    <button type="button" className="google-button mt-2" onClick={handleGoogleLogin}>
                        <span className="google-button__icon">
                            <svg viewBox="0 0 366 372" xmlns="http://www.w3.org/2000/svg"><path d="M125.9 10.2c40.2-13.9 85.3-13.6 125.3 1.1 22.2 8.2 42.5 21 59.9 37.1-5.8 6.3-12.1 12.2-18.1 18.3l-34.2 34.2c-11.3-10.8-25.1-19-40.1-23.6-17.6-5.3-36.6-6.1-54.6-2.2-21 4.5-40.5 15.5-55.6 30.9-12.2 12.3-21.4 27.5-27 43.9-20.3-15.8-40.6-31.5-61-47.3 21.5-43 60.1-76.9 105.4-92.4z" id="Shape" fill="#EA4335"/><path d="M20.6 102.4c20.3 15.8 40.6 31.5 61 47.3-8 23.3-8 49.2 0 72.4-20.3 15.8-40.6 31.6-60.9 47.3C1.9 232.7-3.8 189.6 4.4 149.2c3.3-16.2 8.7-32 16.2-46.8z" id="Shape" fill="#FBBC05"/><path d="M361.7 151.1c5.8 32.7 4.5 66.8-4.7 98.8-8.5 29.3-24.6 56.5-47.1 77.2l-59.1-45.9c19.5-13.1 33.3-34.3 37.2-57.5H186.6c.1-24.2.1-48.4.1-72.6h175z" id="Shape" fill="#4285F4"/><path d="M81.4 222.2c7.8 22.9 22.8 43.2 42.6 57.1 12.4 8.7 26.6 14.9 41.4 17.9 14.6 3 29.7 2.6 44.4.1 14.6-2.6 28.7-7.9 41-16.2l59.1 45.9c-21.3 19.7-48 33.1-76.2 39.6-31.2 7.1-64.2 7.3-95.2-1-24.6-6.5-47.7-18.2-67.6-34.1-20.9-16.6-38.3-38-50.4-62 20.3-15.7 40.6-31.5 60.9-47.3z" fill="#34A853"/></svg>
                        </span>
                        <span className="google-button__text">Sign in with Google</span>
                    </button>
                </div>
                <Link 
                    to='/auth/register'
                    className='link auth__new-account'
                >
                    Create new account
                </Link>
            </form>
        </div>
    )
}
