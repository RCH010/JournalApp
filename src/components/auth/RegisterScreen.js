import React from 'react'
import { Link } from "react-router-dom";
import { useForm } from '../../hooks/useForm';
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';
import validator from 'validator';
import { startRegisterWithEmailPasswordName } from '../../actions/auth';

export const RegisterScreen = () => {

    const dispatch = useDispatch()
    const {msgError} = useSelector( state => state.ui);

    const [formValues, handleInputChange] = useForm({
        name: '',
        email: '',
        password: '',
        password2: '',
    });

    const { name, email, password, password2 } = formValues;

    const handleRegister = (e) => {
        e.preventDefault();
        if(isFormValid()) {
            dispatch(startRegisterWithEmailPasswordName(email,password,name));
        } 
    }

    const isFormValid = () => {
        if(name.trim().length === 0){
            dispatch(setError('Name is required.'));
            return false;
        }
        if (!validator.isEmail(email)){
            dispatch(setError('Email is not valid.'));
            return false;
        }
        if (password !== password2 || password.length <= 5){
            dispatch(setError('Password should be at least 6 characters and match with each other.'));
            return false;
        }
        dispatch(removeError());
        return true;
    }

    

    return (
        <div>
            <h3 className='auth__title'>Register</h3>

            <form onSubmit={handleRegister} className='animate__animated animate__fadeIn animate__faster'>
                {
                    msgError && 
                    (<div className='auth__alert-error'>
                        {msgError}
                    </div>)
                }
                <input 
                    type='text'
                    placeholder='Name'
                    name='name'
                    className='auth__input'
                    autoComplete='off'
                    value={name}
                    onChange={handleInputChange}
                />
                <input 
                    type='email'
                    placeholder='myjournal@gmail.com'
                    name='email'
                    className='auth__input'
                    autoComplete='off'
                    value={email}
                    onChange={handleInputChange}
                />
                <input 
                    type='password'
                    placeholder='Password'
                    name='password'
                    className='auth__input'
                    value={password}
                    onChange={handleInputChange}
                />
                <input 
                    type='password'
                    placeholder='Confirm Password'
                    name='password2'
                    className='auth__input'
                    value={password2}
                    onChange={handleInputChange}
                />

                <button
                    type='submit'
                    className='btn btn-primary btn-block mb-1 mt-1'
                >
                    Register
                </button>
                
                <Link 
                    to='/auth/login'
                    className='link'
                >
                    Already registered?
                </Link>
            </form>
        </div>
    )
}
