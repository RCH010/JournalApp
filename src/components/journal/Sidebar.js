import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { startLogout } from '../../actions/auth';
import { startNewNote } from '../../actions/notes';
import { JournalEntries } from './JournalEntries'

export const Sidebar = () => {

    const dispatch = useDispatch();
    const user = useSelector(state => state.auth)
    const handleLogout = () => {
        dispatch(startLogout());
    }

    const handleNewEntry = () => {
        console.log('do');
        dispatch(startNewNote());
    }
    return (
        <aside className='journal__sidebar'>
            <div className='journal__sidebar-navbar mt-1'>
                <h3 className='journal__sidebar-user-name'>
                    <i className='far fa-moon' />
                    <span> {user.name}</span>
                </h3>

                <button 
                    className='btn' 
                    onClick={handleLogout}
                >
                    Logout
                </button>
            </div>

            <div className='journal__new-entry' onClick={handleNewEntry}>
                <i className='far fa-calendar-plus fa-3x' />
                <p className='mt-1'>
                    New entry
                </p>
            </div>

            <JournalEntries />
        </aside>
    )
}
