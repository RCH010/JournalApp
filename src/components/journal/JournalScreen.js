import React from 'react'
import { useSelector } from 'react-redux'
import { NoteScreen } from '../notes/NoteScreen'
import { NothingSelected } from './NothingSelected'
import { Sidebar } from './Sidebar'

export const JournalScreen = () => {

    const {active} = useSelector(state => state.notes)

    return (
        <div className='journal__main-content'>
            <Sidebar />
            <main>
                {
                    (active) 
                        ? (<NoteScreen />)
                        : (<NothingSelected />)
                }
            </main>
        </div>
    )
}
