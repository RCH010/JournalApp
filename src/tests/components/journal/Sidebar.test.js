import { mount } from "enzyme"
import { Provider } from "react-redux"
import configureStore from 'redux-mock-store';
import thunk from "redux-thunk";
import { startLogout } from "../../../actions/auth";
import { startNewNote } from "../../../actions/notes";
import { Sidebar } from "../../../components/journal/Sidebar"

jest.mock('../../../actions/notes', () => ({
    startNewNote: jest.fn()
}))
jest.mock('../../../actions/auth', () => ({
    startLogout: jest.fn()
}))

const middlewares = [thunk];
const mockStore = configureStore(middlewares);

const initState = {
    auth: {
        uid: '1234ABC',
        name: 'Testing',
        email:'test@tesing.com'
    },
    ui: {
        loading: false,
        msgError: null
    },
    notes: {
        notes: [],
        active: null
    }
};

let store = mockStore(initState)
store.dispatch = jest.fn();

const wrapper = mount(
    <Provider store={store}>
            <Sidebar />
    </Provider>
)

describe('Tests in <Sidebar/>', () => {

    test('should match Sidebar with snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('should call logout when btn pressed', () => {
        wrapper.find('button').prop('onClick')();
        expect( startLogout ).toHaveBeenCalled()

    })

    test('should call startNewNote when btn pressed', () => {
        wrapper.find('.journal__new-entry').prop('onClick')();
        expect(startNewNote).toHaveBeenCalled();
    })
    
    
    
})
