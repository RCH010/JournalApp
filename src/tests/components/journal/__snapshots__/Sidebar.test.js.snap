// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in <Sidebar/> should match Sidebar with snapshot 1`] = `
<aside
  className="journal__sidebar"
>
  <div
    className="journal__sidebar-navbar mt-1"
  >
    <h3
      className="journal__sidebar-user-name"
    >
      <i
        className="far fa-moon"
      />
      <span>
         
        Testing
      </span>
    </h3>
    <button
      className="btn"
      onClick={[Function]}
    >
      Logout
    </button>
  </div>
  <div
    className="journal__new-entry"
    onClick={[Function]}
  >
    <i
      className="far fa-calendar-plus fa-3x"
    />
    <p
      className="mt-1"
    >
      New entry
    </p>
  </div>
  <div
    className="journal__entires"
  />
</aside>
`;
